### PR TITLE
call EL functions based on timestamp

### PR DESCRIPTION
**Pull request by rolfyone**
Signed-off-by: Paul Harris <paul.harris@consensys.net>

## Documentation

- [X] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [X] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: ethereum/executionclient/src/main/java/tech/pegasys/teku/ethereum/executionclient/schema/ExecutionPayloadV2.java [hunk 1]
```diff
@@ -135,4 +135,26 @@ public static List<WithdrawalV1> getWithdrawals(
     }
     return withdrawals;
   }
+
+  @Override
+  public String toString() {
+    final StringBuilder sb = new StringBuilder("ExecutionPayloadV2{");
```
Comment (tbenr) [inline ethereum/executionclient/src/main/java/tech/pegasys/teku/ethereum/executionclient/schema/ExecutionPayloadV2.java:136]: Is there a reason why you're not using `MoreObjects::toStringHelper` like elsewhere?

File: ethereum/executionclient/src/main/java/tech/pegasys/teku/ethereum/executionclient/schema/GetPayloadV2Response.java [hunk 2]
```diff
@@ -33,4 +33,13 @@ public GetPayloadV2Response(
     this.executionPayload = executionPayload;
     this.blockValue = blockValue;
   }
+
+  @Override
+  public String toString() {
+    final StringBuilder sb = new StringBuilder("GetPayloadV2Response{");
```
Comment (tbenr) [inline ethereum/executionclient/src/main/java/tech/pegasys/teku/ethereum/executionclient/schema/GetPayloadV2Response.java:39]: here too

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/executionlayer/ForkChoiceState.java [hunk 3]
```diff
@@ -66,16 +69,22 @@ public boolean isHeadOptimistic() {
...
   public String toString() {
-    return MoreObjects.toStringHelper(this)
-        .add("headBlockRoot", headBlockRoot)
-        .add("headBlockSlot", headBlockSlot)
-        .add("headExecutionBlockHash", headExecutionBlockHash)
-        .add("safeExecutionBlockHash", safeExecutionBlockHash)
-        .add("finalizedExecutionBlockHash", finalizedExecutionBlockHash)
-        .add("isHeadOptimistic", isHeadOptimistic)
-        .toString();
+    final StringBuilder sb = new StringBuilder("ForkChoiceState{");
```
Comment (tbenr) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/executionlayer/ForkChoiceState.java:78]: here too

File: ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/CapellaExecutionClientHandler.java [hunk 4]
```diff
@@ -36,15 +39,38 @@
 public class CapellaExecutionClientHandler extends BellatrixExecutionClientHandler
     implements ExecutionClientHandler {
   private static final Logger LOG = LogManager.getLogger();
+  private Optional<UInt64> firstCapellaSlot = Optional.empty();
```
Comment (tbenr) [inline ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/CapellaExecutionClientHandler.java:42]: I think we could reduce the noise by making this final and non-optional, like:
image

We shouldn't try to instantiate this handler if Capella is not even defined..

File: ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/CapellaExecutionClientHandler.java [hunk 5]
```diff
@@ -75,6 +101,31 @@ public SafeFuture<ExecutionPayloadWithValue> engineGetPayload(
...
+          Optional.of(
+              forkChoiceState
+                  .getGenesisTime()
+                  .get()
+                  .plus(
+                      firstCapellaSlot
+                          .get()
+                          .times(spec.getSecondsPerSlot(firstCapellaSlot.get()))));
+      LOG.trace("Capella timestamp={}", capellaTimestamp.get());
+    }
```
Comment (tbenr) [inline ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/CapellaExecutionClientHandler.java:117]: maybe move it in a dedicated function might clean-up the code here...

**Review by tbenr — COMMENTED** [review]
Provided some comments for small improvements but this is incredibly annoying!

I'm thinking about alternatives. Like having engineForkChoiceUpdated accepting a wrapper of `ForkChoiceState` like `ForkChoiceStateAndCurrentMilestone`.
Then let ForkChoiceNotifier and ForkChoiceUpdateData handle it (ForkChoiceNotifier has access to recentchaindata).

This way we could keep ForkChoice out of the mess...

**Comment by rolfyone** [comment]
thanks for the feedback Enrico 
I'm going to park this PR, and mark draft for now, I think given conversation with Mikhail there's a better way to handle this...
The moreObjects vs.  string buffer thing was just intellij doing its thing, i'll change the setting...
There's basically a huge refactor we can do in this area, remove the need for that timestamp i use currently, but will introduce a requirement to know the slot we're talking about, which i think is a much better position to be in over all.

**Comment by rolfyone** [comment]
This PR is no longer required as the alternate slot approach worked well.

**Pull request closed without merging**